vespalib/src/vespa/vespalib/net/tls/crypto_codec.h

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#pragma once
+
+#include <memory>
+
+namespace vespalib::net::tls {
+
+struct HandshakeResult {
+    // Handshake bytes consumed from peer.
+    size_t bytes_consumed = 0;
+    // Handshake bytes produced that must be sent to the peer.
+    size_t bytes_produced = 0;
+    enum class State {
+        Failed,
+        Done,
+        NeedsMorePeerData
+        // TODO add Closed/Shutdown as own state?
+    };
+    State state = State::Failed;
+
+    bool failed() const noexcept { return (state == State::Failed); }
+    bool done() const noexcept { return (state == State::Done); }
+};
+
+struct EncodeResult {
+    // Plaintext bytes consumed
+    size_t bytes_consumed = 0;
+    // Ciphertext bytes produced that must be sent to the peer
+    size_t bytes_produced = 0;
+    bool failed = true;
+};
+
+struct DecodeResult {
+    // Ciphertext bytes consumed from peer
+    size_t bytes_consumed = 0;
+    // Plaintext bytes produced.
+    size_t bytes_produced = 0;
+    enum class State {
+        Failed,
+        OK,
+        NeedsMorePeerData
```

Comment:
The presence of "give more data pls" return codes is so prevalent in the OpenSSL API that I felt it would be safest to expose it explicitly for now. I think it should actually be possible to call decode() the way you suggest with the code as it is today. Let's see what we need for the integration first, then remove it if we find that it's not actually needed at all 🙂